src/stub_read.s

```diff
@@ -1,47 +1,60 @@
+            .extern _Zp
 
-CHKIN:	      .equlab 0xffc6
-CHRIN:	      .equlab 0xffcf
-READST:	      .equlab 0xffb7
-CLRCHN:       .equlab 0xffcc
+CHKIN:	    .equlab 0xffc6
+CHRIN:	    .equlab 0xffcf
+READST:	    .equlab 0xffb7
+CLRCHN:     .equlab 0xffcc
 
-              .section code
-              .public _Stub_read
-_Stub_read:   ldx     zp:_Zp+0
-	      beq     start$	    ; stdin
-	      cpx     #3	    ; test for stdout/stderr
-	      bcc     eof$
-	      jsr     CHKIN	    ; ensure channel is prepared to read
-start$:	      lda     #0	    ; set file counter = 0
-              sta     zp:_Zp+0
-              sta     zp:_Zp+1
-              ldx     zp:_Zp+5      ; 256 or more?
-              beq     20$           ; no
-              ldy     #0
-10$:          jsr     CHRIN
-              sta     (_Zp+2),y
-              iny
-              bne     10$
-              inc     zp:_Zp+3      ; step 256
-              inc     zp:_Zp+1
-              dec     zp:_Zp+5      ; yes, decrement 256 counter
-              bne     10$
+            .section code
+            .public _Stub_read
+_Stub_read: ldx     zp:_Zp+0
+            beq     start$	    ; stdin
+	        cpx     #3	    ; test for stdout/stderr
+            bcc     eof$
+	        jsr     CHKIN	    ; ensure channel is prepared to read
+start$:	    lda     #0	    ; set file counter = 0
+            sta     zp:_Zp+0
+            sta     zp:_Zp+1
+            lda     zp:_Zp+4
+            ora     zp:_Zp+5
+            beq     30$
+            jsr     CHRIN         ; get first byte
+            tax
+            jsr     READST
+            and     #0x02
+            bne     eof$
+            txa
+            ldy     #0
+            ldx     zp:_Zp+5      ; 256 or more?
+            beq     20$           ; no
+            bne     15$
+10$:        jsr     CHRIN
+15$:        sta     (_Zp+2),y
+            inc     zp:_Zp+0
+            jsr     READST
+            tax
+            bne     30$
+            iny
+            bne     10$
+            inc     zp:_Zp+3      ; step 256
+            inc     zp:_Zp+1
+            dec     zp:_Zp+5      ; yes, decrement 256 counter
+            bne     10$
 
-20$:          ldy     #0
-25$:          cpy     zp:_Zp+4
-              beq     30$
-              jsr     CHRIN
-              sta     (_Zp+2),y
-              inc     zp:_Zp+0
-              iny
-              bne     25$
-30$:	      jsr     READST	    ; check for error
-	      tay
-          jsr     CLRCHN
-          tya
-          tax 
-	      beq     done$
-eof$:	      lda     #-1
-	      sta     zp:_Zp+0
-	      sta     zp:_Zp+1
-done$:	      rts
+25$:        cpy     zp:_Zp+4
+            beq     30$
+            jsr     CHRIN
+20$:        sta     (_Zp+2),y
+            inc     zp:_Zp+0
+            jsr     READST
+            tax
+            bne     30$
+            iny
+            bne     25$
+30$:	    jsr     CLRCHN
+	        beq     done$
+eof$:	    jsr     CLRCHN
+            lda     #0
```

Comment:
It we go to `eof$` I assume there was an error and in that case it should perhaps return `-1`?

Reply:
This one, I had to go around a lot about. EOI (what Commdore calls EOF) is a "normal" condition, so it's special, in that nothing is readable, but it doesn't get it's own POSIX error code.

When -1 is returned, this routine itself is fine, but fread() locks up. I looked through the __lib_fread code, and I didn't see why that would happen. But, I did see this:

              else if (bytes_read == 0)
                {
                  /* We are at the end of the file.  But we may already
                   * have buffered data.  In that case, we will report
                   * the EOF indication later.
                   */

                  break;
  
  fread itself is looking for _Stub_read to return 0 in the event of EOF.
  
  I double-checked with the POSIX standard, and it says to effect the same thing
  https://pubs.opengroup.org/onlinepubs/9699919799/functions/read.html
No data transfer shall occur past the current end-of-file. If the starting position is at or after the end-of-file, 0 shall be returned.

Since that seemed to both agree with __lib_fread, and fix the lockup at the end of file, it seemed safe enough to go with?

Reply:
I realize that EOF has two meanings depending on who you ask.

Reaching the end of file, or EOI as Commodore calls it can be a normal condition. In `stdio.h` the `EOF` symbol is typically an error condition where the error code is stored in `errno`.

Which means that the test for `stdout` and `stderr` in the beginning should be a call to some error label, and set `errno`. Looking in `errno.h` I wonder if `EBADF` would be appropriate. This should return `-1` also.